libs/javalib/worker/src/mill/javalib/zinc/ZincWorkerRpcServer.scala

```diff
@@ -0,0 +1,224 @@
+          auxiliaryClassFileExtensions = msg.auxiliaryClassFileExtensions,
+          compilerBridgeData = clientRequestId
+        )(using msg.ctx, deps)
+      }
+
+      private def docJar(
+          clientRequestId: MillRpcRequestId,
+          msg: ClientToServer.DocJar
+      ): msg.Response =
+        worker.docJar(
+          scalaVersion = msg.scalaVersion,
+          scalaOrganization = msg.scalaOrganization,
+          compilerClasspath = msg.compilerClasspath,
+          scalacPluginClasspath = msg.scalacPluginClasspath,
+          args = msg.args,
+          compilerBridgeData = clientRequestId
+        )
+    }
+  }
+}
+object ZincWorkerRpcServer {
+
+  /**
+   * @param taskDest the task's destination folder.
+   */
+  case class Initialize(taskDest: os.Path, jobs: Int, compileToJar: Boolean, zincLogDebug: Boolean)
+      derives ReadWriter
+
+  sealed trait ReporterMode derives ReadWriter {
+    def reportCachedProblems: Boolean
+  }
+  object ReporterMode {
+    case object NoReporter extends ReporterMode {
+      override def reportCachedProblems: Boolean = false
+    }
+
+    case class Reporter(reportCachedProblems: Boolean, maxErrors: Int) extends ReporterMode
+  }
+
+  sealed trait ClientToServer extends MillRpcMessage derives ReadWriter
+  object ClientToServer {
+    case class CompileJava(
+        upstreamCompileOutput: Seq[CompilationResult],
+        sources: Seq[os.Path],
+        compileClasspath: Seq[os.Path],
+        javacOptions: JavaCompilerOptions,
+        reporterMode: ReporterMode,
+        incrementalCompilation: Boolean,
+        ctx: ZincWorker.InvocationContext
+    ) extends ClientToServer {
+      override type Response = Result[CompilationResult]
+    }
+
+    case class CompileMixed(
+        upstreamCompileOutput: Seq[CompilationResult],
+        sources: Seq[os.Path],
+        compileClasspath: Seq[os.Path],
+        javacOptions: JavaCompilerOptions,
+        scalaVersion: String,
+        scalaOrganization: String,
+        scalacOptions: Seq[String],
+        compilerClasspath: Seq[PathRef],
+        scalacPluginClasspath: Seq[PathRef],
+        reporterMode: ReporterMode,
+        incrementalCompilation: Boolean,
+        auxiliaryClassFileExtensions: Seq[String],
+        ctx: ZincWorker.InvocationContext
+    ) extends ClientToServer {
+      override type Response = Result[CompilationResult]
+    }
+
+    case class DocJar(
+        scalaVersion: String,
+        scalaOrganization: String,
+        compilerClasspath: Seq[PathRef],
+        scalacPluginClasspath: Seq[PathRef],
+        args: Seq[String]
+    ) extends ClientToServer {
+      override type Response = Boolean
+    }
+  }
+
+  sealed trait ServerToClient extends MillRpcMessage derives ReadWriter
+  object ServerToClient {
+    case class AcquireZincCompilerBridge(scalaVersion: String, scalaOrganization: String)
+        extends ServerToClient {
+      override type Response = ZincCompilerBridge.AcquireResult[os.Path]
+    }
```

Comment:
Could we move the downloading of the compiler bridge to `JvmWorkerImpl`, before we begin RPCing to `ZincWorker`? We already have all the necessary data (`scalaVersion` and `scalaOrganization`) pretty early, so moving it to `JvmWorkerImpl` would save us one back-and-forth and simplify the RPC data flow

Reply:
I don't think we should touch that logic now. I tried doing as little as possible with the actual compilation logic and this seems like another, unrelated, refactoring.